api/src/services/users.js

```diff
@@ -432,51 +481,216 @@ const missingFields = data => {
     settings._id = userID;
     return settings;
   });
 };
 
+const shouldEnableTokenLogin = data => {
+  const tokenLoginConfig = config.get('token_login');
+  const isValidConfig = tokenLoginConfig &&
+                        (tokenLoginConfig.translation_key || tokenLoginConfig.message) &&
+                        tokenLoginConfig.app_url;
+  if (!isValidConfig) {
+    return false;
+  }
+
+  return data.token_login;
+};
+const shouldDisableTokenLogin = data => data.token_login === false;
+
+const getHash = string => crypto.createHash('sha1').update(string, 'utf8').digest('hex');
+
+/**
+ * Clears pending tasks in the currently assigned `token_login_sms` doc
+ * Used to either disable or refresh token-login
+ *
+ * @param {Object} user - the _users doc
+ * @param {Object} user.token_login - token-login information
+ * @param {Boolean} user.token_login.active - whether the token-login is active or not
+ * @param {String} user.token_login.doc_id - the id of the current `token_login_sms`
+ * @returns {Promise}
+ */
+const clearOldLoginSms = (user) => {
+  if (!user.token_login || !user.token_login.active || !user.token_login.doc_id) {
+    return Promise.resolve();
+  }
+
+  return db.medic
+    .get(user.token_login.doc_id)
+    .then(doc => {
+      if (!doc || !doc.tasks) {
+        return;
+      }
+
+      const pendingTasks = doc.tasks.filter(task => task.state === 'pending');
+      if (!pendingTasks.length) {
+        return;
+      }
+
+      pendingTasks.forEach(task => taskUtils.setTaskState(task, 'cleared'));
+      return db.medic.put(doc);
+    })
+    .catch(err => {
+      if (err.status === 404) {
+        return;
+      }
+      throw err;
+    });
+};
+
+/**
+ * Generates a doc of type `token_login_sms` that contains the two tasks required for token login: one reserved
+ * for the URL and one containing instructions for the user.
+ * If the user already had token-login activated, the pending tasks of the existent `token_login_sms` doc are cleared.
+ * @param {Object} user - the _users document
+ * @param {Object} userSettings - the user-settings document from the main database
+ * @param {String} token - the randomly generated token
+ * @param {String} userHash - sha1 hash of the user's document id
+ * @param {Object} tokenLoginConfig - the token_login config section that contains the translation_key used to generate
+ *                                    the instructions sms message
+ * @returns {Promise<Object>} - returns the result of saving the new token_login_sms document
+ */
+const generateLoginSms = (user, userSettings, token, userHash, tokenLoginConfig) => {
+  return clearOldLoginSms(user).then(() => {
+    const doc = {
+      type: 'token_login_sms',
+      reported_date: new Date().getTime(),
+      user: user._id,
+      tasks: [],
+    };
+    const userHash = getHash(user._id);
+    const url = `${tokenLoginConfig.app_url.replace(/\/+$/, '')}/medic/login/token/${token}/${userHash}`;
```

Comment:
I'd like to get rid of the userHash if possible, to save on SMS length but also to reduce complexity. Isn't the token enough to find the user?

Reply:
If I dump the view, the hash (that won't be hash anymore) will be necessary to find the user.